persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/DelegatingAuthorizer.java

```diff
@@ -0,0 +1,183 @@
+import io.stargate.auth.entity.ResourceKind;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import org.apache.cassandra.auth.AuthenticatedUser;
+import org.apache.cassandra.auth.CassandraAuthorizer;
+import org.apache.cassandra.auth.DataResource;
+import org.apache.cassandra.auth.FunctionResource;
+import org.apache.cassandra.auth.IResource;
+import org.apache.cassandra.auth.Permission;
+import org.apache.cassandra.auth.RoleResource;
+import org.apache.cassandra.exceptions.RequestExecutionException;
+import org.apache.cassandra.exceptions.RequestValidationException;
+import org.apache.cassandra.stargate.exceptions.AuthenticationException;
+
+public class DelegatingAuthorizer extends CassandraAuthorizer {
+
+  private static final long PROCESSING_TIMEOUT_MILLIS =
+      Long.getLong("stargate.authorization.processing.timeout.millis", 5 * 60 * 1000);
+
+  private AuthorizationProcessor authProcessor;
+
+  public void setProcessor(AuthorizationProcessor processor) {
+    authProcessor = processor;
+  }
+
+  @Override
+  public void grant(
+      AuthenticatedUser performer,
+      Set<Permission> permissions,
+      IResource resource,
+      RoleResource grantee)
+      throws RequestValidationException, RequestExecutionException {
+    if (authProcessor == null) {
+      super.grant(performer, permissions, resource, grantee);
+      return;
+    }
+
+    CompletionStage<Void> stage =
+        authProcessor.addPermissions(
+            ImmutableActor.of(performer.getName()),
+            AuthorizationOutcome.ALLOW,
+            PermissionKind.ACCESS,
+            permissions(permissions),
+            resource(resource),
+            role(grantee));
+
+    get(stage);
+  }
+
+  @Override
+  public void revoke(
+      AuthenticatedUser performer,
+      Set<Permission> permissions,
+      IResource resource,
+      RoleResource revokee)
+      throws RequestValidationException, RequestExecutionException {
+    if (authProcessor == null) {
+      super.revoke(performer, permissions, resource, revokee);
+      return;
+    }
+
+    CompletionStage<Void> stage =
+        authProcessor.removePermissions(
+            ImmutableActor.of(performer.getName()),
+            AuthorizationOutcome.ALLOW,
+            PermissionKind.ACCESS,
+            permissions(permissions),
+            resource(resource),
+            role(revokee));
+
+    get(stage);
+  }
+
+  private static void get(CompletionStage<Void> stage) {
+    try {
+      // wait for completion since the calling API is synchronous
+      stage.toCompletableFuture().get(PROCESSING_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof org.apache.cassandra.stargate.exceptions.RequestValidationException) {
```

Comment:
is it possible that `cause == null`?

Reply:
yes, and the `instanceof` result will be `false` in that case.